### PR TITLE
Activation Hessian computation runtime optimization

### PR DESCRIPTION
**Pull request by ofirgo**
## Pull Request Description:
Improve Activation Hessian computation runtime for GPTQ and Mixed precision with the following optimizations:
- Enable batch computation.
- Enable computation on a set of nodes (instead of a single node only).
- Other minor loop and implementation modifications.

Major design remarks:
- TraceHessianRequest receives a list of target nodes (target_nodes) instead of a single BaseNode.
- HessianInfoService produces a batch of samples for each computation iteration. We also fixed a bug where the HessianInfoService representative generator would have been initiated for each iteration instead of running on the generator from end-to-end.
- - The output of the "fetch" call of HessianInfoService is of the structure: List (per target nodes) of List (per image) of Hessian approximations (tensor).
- - The service's cache mechanism still saves results per request for a single node. We split and construct requests per-node to save and retrieve results for a certain node.
- - Batch computation requires keeping track of remaining samples from a given representative dataset batch, in case the requested Hessian computation batch is smaller (we don't want to "throw" samples away).
- - In addition, we assume that the Hessians computation batch_size is <= to the representative dataset batch size.
- Weights Hessian computation is still limited to a single image and per-node computation.
- Hessian results tensor includes a batch dimension.
- Default values for the number of samples and number of iterations for Hessians computation for GPTQ and Mixed precision have been modified.

## Checklist before requesting a review:
- [x] I set the appropriate labels on the pull request.
- [x] I have added/updated the release note draft (if necessary).
- [x] I have updated the documentation to reflect my changes (if necessary).
- [x] All function and files are well documented. 
- [x] All function and classes have type hints.
- [x] There is a licenses in all file.
- [x] The function and variable names are informative. 
- [x] I have checked for code duplications.
- [x] I have added new unittest (if necessary).

### CODE REVIEW

File: model_compression_toolkit/core/common/hessian/hessian_info_service.py [hunk 1]
```diff
@@ -92,115 +140,233 @@ def count_saved_info_of_request(self, hessian_request:TraceHessianRequest) -> in
...
+                                f"but found only {len(res_for_node)}.")
 
+            res_for_node = res_for_node[:required_size]
 
+            collected_results.append(res_for_node)
+
+        return collected_results
+
+    @staticmethod
+    def _construct_single_node_request(mode: HessianMode, granularity: HessianInfoGranularity, target_nodes
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/core/common/hessian/hessian_info_service.py:344]: I usually prefer not to fill the type hints with "Any" where it is restricted to some specific type (which I just can't specify due to imports).
In this case, the target_nodes must be a List, so I'm adding "List" without specifying the type of elements in the list.

File: model_compression_toolkit/core/keras/hessian/activation_trace_hessian_calculator_keras.py [hunk 2]
```diff
@@ -82,88 +82,74 @@ def compute(self) -> List[float]:
...
-                    for j in range(self.num_iterations_for_approximation):  # Approximation iterations
-                        # Getting a random vector with normal distribution
-                        v = tf.random.normal(shape=output.shape, dtype=output.dtype)
-                        f_v = tf.reduce_sum(v * output)
+                ipts_hessian_trace_approx = [tf.Variable([0.0], dtype=tf.float32, trainable=True)
+                                             for _ in range(len(target_activation_tensors))]
 
+                # Loop through each interest point activation tensor
+                prev_mean_results = None
+                for j in tqdm(range(20)):  # Approximation iterations
```
[the remark this answers is not in the document]
Comment (ofirgo) [reply model_compression_toolkit/core/keras/hessian/activation_trace_hessian_calculator_keras.py:107]: We should, it is just a leftover from debugging I guess... good catch!

**Review by reuvenperetz — APPROVED** [review]

File: model_compression_toolkit/core/common/mixed_precision/mixed_precision_quantization_config.py [hunk 3]
```diff
@@ -44,6 +45,7 @@ def __init__(self,
             norm_scores (bool): Whether to normalize the returned scores for the weighted distance metric (to get values between 0 and 1).
             refine_mp_solution (bool): Whether to try to improve the final mixed-precision configuration using a greedy algorithm that searches layers to increase their bit-width, or not.
             metric_normalization_threshold (float): A threshold for checking the mixed precision distance metric values, In case of values larger than this threshold, the metric will be scaled to prevent numerical issues.
+            hessian_batch_size (int): The Hessian computation batch size. used only if using mixed precision with Hessian-based objective.
```
Comment (reuvenperetz) [inline model_compression_toolkit/core/common/mixed_precision/mixed_precision_quantization_config.py:48]: Used...

**Review by reuvenperetz — APPROVED** [review]
